File: sdk-py/noether_sdk/sub/tx.py
from ..errors import AuthError
from ..models import SubmittedTx
from ..transport import Credentials, Transport


class TxApi:
    def __init__(self, transport: Transport, credentials: Credentials | None) -> None:
        self._transport = transport
        self._credentials = credentials

    async def submit(self, *, signed_xdr: str, poll_timeout_ms: int | None = None) -> SubmittedTx:
        if self._credentials is None:
            raise AuthError(
                "tx.submit requires an authenticated client",
                status=401,
                body=None,
                url="/v1/tx/submit",
            )
        payload: dict[str, object] = {"signedXdr": signed_xdr}
        if poll_timeout_ms is not None:
            payload["pollTimeoutMs"] = poll_timeout_ms
        body = await self._transport.request(
            "POST",
            "/v1/tx/submit",
            json=payload,
            credentials=self._credentials,
        )
        return SubmittedTx.model_validate(body)

File: sdk-ts/src/sub/tx.ts
import type { Credentials, Transport } from '../transport.js';

export interface SubmitRequest {
  signedXdr: string;
  pollTimeoutMs?: number;
}

export interface SubmittedTx {
  hash: string;
  status: 'SUCCESS' | 'FAILED' | 'PENDING' | 'NOT_FOUND';
  ledger?: number;
  /** Decoded Error(Contract, #N) when the contract reverted; null otherwise. */
  contractError?: { code: number; name?: string; message?: string } | null;
  /**
   * Decoded non-contract host error (e.g. { type: 'storage', code: 'exceeded_limit' }
   * — the state moved between simulation and apply); null otherwise.
   */
  hostError?: { type: string; code: string } | null;
}

export class TxApi {
  constructor(private readonly transport: Transport, private readonly credentials: Credentials | null) {}

  async submit(request: SubmitRequest): Promise<SubmittedTx> {
    if (!this.credentials) throw new Error('tx.submit requires an authenticated client');
    return this.transport.request<SubmittedTx>({
      method: 'POST',
      path: '/v1/tx/submit',
      body: request,
      credentials: this.credentials,
    });
  }
}
